nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java

```diff
@@ -100,101 +96,61 @@ private ProvenanceEventRecord getRecord(final Document d, final RecordReader rea
             }
         }
 
-        if ( record == null ) {
-            throw new IOException("Failed to find Provenance Event " + d);
-        } else {
-            return record;
+        if (record == null) {
+            logger.warn("Failed to read Provenance Event for '" + d + "'. The event file may be missing or corrupted");
         }
-    }
 
+        return record;
+    }
 
     public Set<ProvenanceEventRecord> read(final List<Document> docs, final Collection<Path> allProvenanceLogFiles,
-        final AtomicInteger retrievalCount, final int maxResults, final int maxAttributeChars) throws IOException {
-        if (retrievalCount.get() >= maxResults) {
-            return Collections.emptySet();
-        }
-
-        LuceneUtil.sortDocsForRetrieval(docs);
-
-        RecordReader reader = null;
-        String lastStorageFilename = null;
-        final Set<ProvenanceEventRecord> matchingRecords = new LinkedHashSet<>();
+            final AtomicInteger retrievalCount, final int maxResults, final int maxAttributeChars) throws IOException {
 
         final long start = System.nanoTime();
-        int logFileCount = 0;
-
-        final Set<String> storageFilesToSkip = new HashSet<>();
-        int eventsReadThisFile = 0;
 
-        try {
-            for (final Document d : docs) {
-                final String storageFilename = d.getField(FieldNames.STORAGE_FILENAME).stringValue();
-                if ( storageFilesToSkip.contains(storageFilename) ) {
-                    continue;
-                }
-
-                try {
-                    if (reader != null && storageFilename.equals(lastStorageFilename)) {
-                        matchingRecords.add(getRecord(d, reader));
-                        eventsReadThisFile++;
-
-                        if ( retrievalCount.incrementAndGet() >= maxResults ) {
-                            break;
-                        }
-                    } else {
-                        logger.debug("Opening log file {}", storageFilename);
-
-                        logFileCount++;
-                        if (reader != null) {
-                            reader.close();
-                        }
+        Set<ProvenanceEventRecord> matchingRecords = new LinkedHashSet<>();
+        if (retrievalCount.get() >= maxResults) {
+            return matchingRecords;
```

Comment:
We def need to be careful in here.  Performance and implications on GC in general are critical in anything prov related.